### ISSUE TITLE
useNormalTexture wrap mode not working

### ISSUE THREAD

**Issue by ffdead** [pmndrs/drei#621]
- `@react-three/drei` version: 7.12.5

### Problem description:

After changing texture `repeat`, the texture is wrapped using the default `THREE.ClampToEdgeWrapping` instead of `THREE.RepeatWrapping`

This seems to be caused by 742a067392f80f57d836447c46841b001b2e2471 since textures are now premtively uploaded to the GPU.

This probably breaks all current usages of `useTexture` where the texture is changed before rendering. 

### Relevant code:

```js
const [normalMapPrimary] = useNormalTexture(
    42, // index of the normal textur
    // second argument is texture attributes
    {
      repeat: [1.5, 1.5]
    }
);
```

### Suggested solution:

1. Change useNormalTexture to set `normalTexture.needsUpdate = true;` to re-upload it to the GPU
2. Revert the change in `useTexture` as the user might want to change the texture, or want to control when that initial upload jank happens (or at least make it optional somehow to avoid double uploads) **(RECOMMENDED)**

**Comment by joshuaellis** [comment]
I think the opt-out solution is probably more than ideal as I suspect if `three` are doing opt-in automatically (may be wrong there) it's because it's more useful to first comers as opposed to experienced devs. 

@drcmda you made the change, What do you think?

**Comment by ffdead** [comment]
In my experience `initTexture` needs to be combined with `compile` and traversing the tree to make all objects visible in order to avoid scroll jank. But I'm happy to keep it as long as we can opt-out.

Please see the suggested fix in #629

**Comment by drcmda** [comment]
@ffdead could you quickly detail how the texture is changed? because the init runs inside a useEffect. that should leave the user time to either mutate it in useMemo or useLayoutEffect. what makes it break eventually?

@joshuaellis the issue was that in threejs the texture upload take a long time. and it doesn't event happen in compile. so you have some objects that aren't in view, but when they are they cause the mainthread to freeze. i asked them to please fix compile but they didn't want to. i then asked if initializing textures is safe and they seemed to affirm.

**Comment by ffdead** [comment]
> @ffdead could you quickly detail how the texture is changed? because the init runs inside a useEffect. that should leave the user time to either mutate it in useMemo or useLayoutEffect. what makes it break eventually?

This is the code use `useNormalTexture`

```
const normalTexture = useTexture(url, { init: false }) as Texture

  React.useEffect(() => {
    if (!normalTexture) return
    normalTexture.wrapS = normalTexture.wrapT = RepeatWrapping
    normalTexture.repeat = new Vector2(repeat[0], repeat[1])
    normalTexture.offset = new Vector2(offset[0], offset[1])
    normalTexture.anisotropy = anisotropy
  }, [normalTexture, anisotropy, repeat, offset])
```

Changing this to `React.useLayoutEffect` seems to solve this issue and is better than `needsUpdate` as it would cause the texture to be uploaded twice to the GPU.

Your argumentation makes sense to me. If people have a usecase that requires to bypass `initTexture` completely they can use `useLoader` 👍 

Will change the PR to make use of `React.useLayoutEffect` instead, thanks!

FYI I never had perfect results relying on `initTeture` and `compile`. In our scroll-rig we traverse the tree, make all objects visible and pre-render the scene once to init all textures and materials. I think that is the old-school proven way, but would be nice to rely on using the API.

**Comment by drcmda** [comment]
yes, that explains it, thank you! useEffect is def wrong here. you absolutely must call needsUpdate if you change texture in uE bc even w/o my commit the threejs could have flushed it before you, it's a race condition.

> Your argumentation makes sense to me. If people have a usecase that requires to bypass initTexture completely they can use useLoader 👍

right, imo we don't need the extra config object on useTexture. useLoader(TextureLoader, url) should be fine in the rare situation in which that's called for.

> FYI I never had perfect results relying on initTeture and compile. In our scroll-rig we traverse the tree, make all objects visible and pre-render the scene once to init all textures and materials. I think that is the old-school proven way, but would be nice to rely on using the API.

ha! you'll love this: https://github.com/pmndrs/drei/blob/master/src/core/Preload.tsx#L28-L38

the new Preload component uses a cubecamera that acts as an eye of sauron, it sees everything. this is all so silly, i asked them to pretty please fix compile but as it is they won't. but useTexture and `<Preload all />` work pretty good.

**Comment by ffdead** [comment]
> ha! you'll love this: https://github.com/pmndrs/drei/blob/master/src/core/Preload.tsx#L28-L38
> 
> the new Preload component uses a cubecamera that acts as an eye of sauron, it sees everything. this is all so silly, i asked them to pretty please fix compile for that as well, but as it is they won't. but useTexture and `<Preload all />` work pretty good.

Brilliant!

**Comment by joshuaellis** [comment]
resolved by #630
